### ISSUE TITLE
Parallels Tools not installed for RHEL

### ISSUE THREAD

**Issue by stephenulmer** [Parallels/packer-examples#48]
Let me start with: I am not Packer fluent, just barely literate.

When building a box with `redhat/build.pkr.hcl`, the resulting image doesn't have Parallels Tools installed, despite running `scripts/rhel/base/parallels.sh`.

I believe that this is due to a race condition, introduced by `scripts/rhel/base/init.sh`.

Most of the provisioners run in a list in the *same* provisioner block. `init.sh` is the first one to run, and ends with a `reboot` command. I separated all of these provisioners, and eventually added breakpoints, and discovered that the reboot takes some time to take effect. So the *next* provisoning script, `sshd.sh`, runs and completes, but the `parallels.sh` gets interrupted while preparing to install prerequisites. The next script, `parallels_folders.sh`, runs after the box reboots.

I have a version of `redhat/build.pkr.hcl` in which I've separated the provisoners into three groups, and had the second group wait before executing. This makes sure that the box starts booting before Packer *tries* to run the `parallels.sh` script. This doesn't eliminate the problem, but seems to make it *much* less likely.

Are the maintainers willing to accept this half-step-fix for the problem? I can do more restructuring of the provisioners, but that would make them less like the other platforms, and I didn't want to make a mess without asking about direction...

Thank you!

**Comment by cjlapao** [comment]
Hi @stephenulmer, 

Thanks for raising this. We noticed something similar in another example we are working on. While this was Ubuntu, it seemed to be a similar issue. We came up with a similar solution. Let me try the solution we used with that one and see if this works with RedHat, and then I'll get back to you.

Give me a couple of days to run some testing

**Comment by cjlapao** [comment]
@stephenulmer would it be possible to test this pr? 

#49 

We moved the installation of the tools to be the initial step and made the machine reboot making sure there is no concurrency there. We run 5 tests and all build the VM correctly using PD 19.4.1

let me know if this helps and we can merge the code

**Comment by stephenulmer** [comment]
I *think* that the current PR just moves the problem around.

I'll make a PR that restructures the provisioners. You can choose to use it or not, but a counter proposal will be the easiest way to describe my concerns.

**Comment by stephenulmer** [comment]
Okay, I added PR #50, which basically refactors the current provisioners for RHEL.

It was tested with/without Red Hat credentials, with/without making a Vagrant box, et al.

I didn't submit it initially because it changes quite a bit, and I didn't want to just show up with a refactoring of something I didn't know if anyone else thought was broken...

There are still some inefficiencies in a few places, but I don't know how they got that way and there may be a purpose I'm not seeing. Maybe there is intentional similarity to another platform, for example.

At any rate, please let me know what you think.

**Comment by cjlapao** [comment]
@stephenulmer 

Thank you for that. I don't see a problem with it. I just had a comment on a behaviour we see on other Linux distributions when installing PD Tools. Please have a look at my comment, and if you have tested that scenario and it works, I will approve and merge the changes.

**Comment by stephenulmer** [comment]
I replied on the PR, which may not have been the right thing to do? Please let me know what you think about what I said about the "desktop" items, and I'll resolve accordingly. Also, I haven't had a chance to test cut-and-paste yet at all. I will do that soon. I'll test before and after my changes.

**Comment by cjlapao** [comment]
@stephenulmer I agree with your comment and let's go with what you propose, I also believe it is the right thing to do

**Comment by stephenulmer** [comment]
**Cut and paste works with PR #50. I think we should merge it as-is.**

I was preparing to update the PR to remove `desktop.sh`, and I noticed that even though everything needed for the desktop got installed by kickstart, the default-target was set to multi-user.target, rather than graphical.target. This can be changed the first time someone logs in, and it works afterward, so everything is correctly installed.

That setting should have been set correctly when the desktop packages were installed in the kickstart -- it is in every other kickstart file I've got that installs a desktop. Right now `desktop.sh` changes the default target, so we can leave it in until I figure out why the installer isn't doing that. In the worst case, I can change the default-target with an inline provisioner when the `install_desktop` variable is set (still removing the `desktop.sh` script), but not understanding this has me annoyed...

I'll take a look at what kickstart is doing and open another issue or PR to make **one** way to install the desktop.
